src/Engine/ProtoCore/DSASM/HeapObjects.cs

```diff
@@ -0,0 +1,659 @@
+                    StackValue value = pair.Value;
+                    StackValue coercedValue = TypeSystem.Coerce(value, type, runtimeCore);
+
+                    dict[key] = coercedValue;
+                }
+            }
+
+            var svArray = heap.AllocateArray(elements);
+            var array = heap.ToHeapObject<DSArray>(svArray);
+            array.Dict = dict;
+            return svArray;
+        }
+
+        /// <summary>
+        /// array[index] = value. The array will be expanded if necessary.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public StackValue SetValueForIndex(int index, StackValue value, RuntimeCore runtimeCore)
+        {
+            index = ExpandByAcessingAt(index);
+            if (index < 0)
+                index += VisibleSize;
+
+            if (index >= VisibleSize || index < 0)
+            {
+                if (runtimeCore != null)
+                    runtimeCore.RuntimeStatus.LogWarning(WarningID.kOverIndexing, Resources.kArrayOverIndexed);
+                return StackValue.Null;
+            }
+
+            StackValue oldValue = GetItemAt(index);
+            SetItemAt(index, value);
+            return oldValue;
+        }
+
+        /// <summary>
+        /// array[index] = value. Here index can be any type. 
+        /// 
+        /// Note this function doesn't support the replication of array indexing.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public StackValue SetValueForIndex(StackValue index, StackValue value, RuntimeCore runtimeCore)
+        {
+            if (index.IsNumeric)
+            {
+                index = index.ToInteger();
+                return SetValueForIndex((int)index.opdata, value, runtimeCore);
+            }
+            else
+            {
+                StackValue oldValue;
+                if (Dict.TryGetValue(index, out oldValue))
+                {
+                    oldValue = StackValue.Null;
+                }
+                Dict[index] = value;
+
+                return oldValue;
+            }
+        }
+
+        /// <summary>
+        /// array[index1][index2][...][indexN] = value, and
+        /// indices = {index1, index2, ..., indexN}
+        ///
+        /// Note this function doesn't support the replication of array indexing.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <param name="value"></param>
+        /// <param name="core"></param>
+        /// <returns></returns>
+        public StackValue SetValueForIndices(StackValue[] indices, StackValue value, RuntimeCore runtimeCore)
+        {
+            DSArray array = this;
+            for (int i = 0; i < indices.Length - 1; ++i)
```

Comment:
Just want to make sure this is intended, because the last element won't be reached by this `for` loop.